File: scripts/stitch-videos.ts
// Python script for video stitching using FFmpeg
// This would be executed server-side to combine all video clips

const stitchVideosScript = `
import subprocess
import json
import sys
import os
from pathlib import Path

def stitch_videos(video_urls: list, audio_url: str, output_path: str):
    """
    Stitch multiple video clips together with audio using FFmpeg
    """
    print(f"[v0] Stitching {len(video_urls)} videos together")
    
    # Download all videos
    video_files = []
    for i, url in enumerate(video_urls):
        video_file = f"video_{i}.mp4"
        subprocess.run([
            "curl", "-o", video_file, url
        ], check=True)
        video_files.append(video_file)
    
    # Download audio
    audio_file = "audio.mp3"
    subprocess.run([
        "curl", "-o", audio_file, audio_url
    ], check=True)
    
    # Create concat file
    concat_file = "concat.txt"
    with open(concat_file, "w") as f:
        for video_file in video_files:
            f.write(f"file '{video_file}'\\n")
    
    # Concatenate videos
    temp_output = "temp_output.mp4"
    subprocess.run([
        "ffmpeg",
        "-f", "concat",
        "-safe", "0",
        "-i", concat_file,
        "-c", "copy",
        temp_output
    ], check=True)
    
    # Add audio
    subprocess.run([
        "ffmpeg",
        "-i", temp_output,
        "-i", audio_file,
        "-c:v", "copy",
        "-c:a", "aac",
        "-shortest",
        output_path
    ], check=True)
    
    # Cleanup
    for video_file in video_files:
        os.remove(video_file)
    os.remove(audio_file)
    os.remove(concat_file)
    os.remove(temp_output)
    
    print(f"[v0] Video stitching complete: {output_path}")
    return output_path

if __name__ == "__main__":
    data = json.loads(sys.argv[1])
    result = stitch_videos(data["videoUrls"], data["audioUrl"], data["outputPath"])
    print(json.dumps({"output": result}))
`

export default stitchVideosScript

File: scripts/stitch-videos.py
import sys
import json
import subprocess
import tempfile
import os
from urllib.request import urlretrieve

def download_file(url, filepath):
    """Download a file from URL to filepath"""
    print(f"[v0] Downloading {url}")
    urlretrieve(url, filepath)
    return filepath

def stitch_videos_with_audio(video_urls, audio_url, output_filename):
    """
    Stitch multiple video clips together and add audio track
    """
    try:
        with tempfile.TemporaryDirectory() as temp_dir:
            # Download all videos
            video_files = []
            for i, url in enumerate(video_urls):
                video_path = os.path.join(temp_dir, f"scene_{i:03d}.mp4")
                download_file(url, video_path)
                video_files.append(video_path)
            
            # Download audio
            audio_path = os.path.join(temp_dir, "audio.mp3")
            download_file(audio_url, audio_path)
            
            # Create concat file for FFmpeg
            concat_file = os.path.join(temp_dir, "concat.txt")
            with open(concat_file, 'w') as f:
                for video_file in video_files:
                    f.write(f"file '{video_file}'\n")
            
            # Concatenate videos without audio
            temp_video = os.path.join(temp_dir, "concatenated.mp4")
            concat_cmd = [
                'ffmpeg',
                '-f', 'concat',
                '-safe', '0',
                '-i', concat_file,
                '-c', 'copy',
                temp_video
            ]
            
            print("[v0] Concatenating video clips...")
            subprocess.run(concat_cmd, check=True, capture_output=True)
            
            # Add audio track to concatenated video
            output_path = os.path.join(temp_dir, output_filename)
            audio_cmd = [
                'ffmpeg',
                '-i', temp_video,
                '-i', audio_path,
                '-c:v', 'copy',
                '-c:a', 'aac',
                '-shortest',  # Match the shortest stream (video or audio)
                output_path
            ]
            
            print("[v0] Adding audio track...")
            subprocess.run(audio_cmd, check=True, capture_output=True)
            
            # Read the final video file
            with open(output_path, 'rb') as f:
                video_data = f.read()
            
            return {
                'success': True,
                'output_path': output_path,
                'size': len(video_data)
            }
            
    except subprocess.CalledProcessError as e:
        return {
            'success': False,
            'error': f'FFmpeg error: {e.stderr.decode() if e.stderr else str(e)}'
        }
    except Exception as e:
        return {
            'success': False,
            'error': str(e)
        }

if __name__ == "__main__":
    # Read input from stdin
    input_data = json.loads(sys.stdin.read())
    
    video_urls = input_data.get('videoUrls', [])
    audio_url = input_data.get('audioUrl')
    output_filename = input_data.get('outputFilename', 'final-video.mp4')
    
    result = stitch_videos_with_audio(video_urls, audio_url, output_filename)
    print(json.dumps(result))
